src/server/master_server/chunk_server_manager.h

```diff
@@ -54,45 +137,132 @@ class ChunkServerManager {
+  // allocations. Note:
+  // 1. The specified chunk_handle is now mapped to the returned locations.
+  // 2. Will return empty set of locations, if there is no chunk server
+  // available for chunk storage. e.g. When chunkservers don't have enough disk
+  // space left or there are no chunk servers registered (i.e. servers down).
+  // 3. If allocate is called for an already allocated chunk, we will return the
+  // previously allocated locations and not do new allocation.
+  ChunkServerLocationThreadSafeFlatSet AllocateChunkServer(
+      const std::string& chunk_handle, const ushort& requested_servers_count);
 
   // Register the ChunkServer with the manager.
   // Manager can now decide to select it for chunk storage.
-  // Returns false if ChunkServer is already registered.
+  // If it's an exisiting server with allocated chunk handles, the chunk handles
+  // will be updated to reflect this server location and future query of the
+  // chunk handle location will show this server location. Threadsafe and allows
+  // concurrent registration. Returns false if ChunkServer is already
+  // registered. Could be because another chunk server is already using that
+  // server location and we only allow unique locations.
   bool RegisterChunkServer(
       const std::shared_ptr<protos::ChunkServer> chunk_server);
 
   // Unregister the ChunkServer with the manager.
   // Manager no longer knows about this server and won't be selected for chunk
-  // storage.
+  // storage. This also removes the chunk server from the locations for
+  // previously allocated chunk handles. When these chunk handles locations are
+  // queried, it won't show this location. After this, the server location can
+  // be used to register the same chunk server again or another chunk server.
   void UnRegisterChunkServer(
       const protos::ChunkServerLocation& server_location);
 
-  // Unregister all the ChunkServers.
-  // There would be no ChunkServer for chunk storage until they are registered.
+  // This is used for testing only. Unregister all the ChunkServers. Not thread
+  // safe on purpose. There would be no ChunkServer for chunk storage until they
+  // are registered.
+  // TODO(bmokutub): Consider making this private and making the unit test class
+  // a friend.
   void UnRegisterAllChunkServers();
 
-  // Returns the ChunkServer for the specified location. Returns nullptr if not
-  // registered.
-  std::shared_ptr<protos::ChunkServer> GetChunkServer(
-      const protos::ChunkServerLocation& server_location);
+  // Returns the ChunkServer for the specified location and the chunkserver
+  // mutex for threadsafe access. Returns a constant so it can't be modified
+  // outside the manager. Returns nullptr if not registered. This is used by the
+  // ChunkServerManager services only.
+  // TODO(bmokutub): Consider making this private and making the service class
+  // a friend.
+  std::pair<std::shared_ptr<const protos::ChunkServer>,
```

Comment:
I think we should just return const protos::ChunkServer here, which is a copy of the chunk server information. 

protos::ChunkServer is not a sophisticated type so copy operation is inexpensive. The proto buffer generates code for copy constructor and we can just use that. 

That way, we don' have to worry about thread-safety for the return items, and won't have to expose locks.

Reply:
haha, very true. I got carried away here. So this method is being used by the chunkservermgr service and was initially planning to do update stuff there that will require lock but later didn't end up doing it and came back here to add const instead of just removing it. Yeah correct will just return a copy. Thanks.

Reply:
Yes sounds good, let's just return a copy, making both of our lives easier :)